woocommerce/woocommerce-ios
# Products tab: fix earlier draft products appearing at the top when sorting by "Newest to Oldest"

Fixes #2727 

The PR has ~2700 diffs 😱 However, the actual changes of this PR are pretty small (under 200 diffs), just a lot of diffs in the Core Data mapping model and model version.

- The changes in `Model 33.xcdatamodel/contents` are in https://github.com/woocommerce/woocommerce-ios/commit/d9c5aa0512525c9fc82e5a90ed16679b811ff699#diff-c2640637c23ff4b62d2a9dc19c9290438213cab88f068214fb49c9d647b5d616
- The only change I made in `WooCommerceModelV32toV33.xcmappingmodel/xcmapping.xml`:
<img width="691" alt="Screen Shot 2020-10-21 at 11 14 19 AM" src="https://user-images.githubusercontent.com/1945542/96680255-bc076700-13a7-11eb-9f30-6b3980ef1e35.png">


## Changes

- Created a new model version 33 with a mapping model that sets the new non-optional `Product.date` to the previous `Product.dateCreated`, plus a test case in `MigrationTests`.
- Added `date: Date` to `Product` model in Networking and Storage layers, updated unit tests. In `Networking.Product`, the new `date` property is set to `dateModified` for a draft product and `dateCreated` otherwise.
- In `ResultsController+SortProducts`, updated `sortDescriptors` for `dateAscending` and `dateDescending` cases to use the new `date` attribute instead of `dateCreated`. Added a test case in `ResultsController_StorageProductTests` that would fail without changes in this PR.

## Testing

Note: this bug is only reproducible on certain sites where a draft product has `null` creation date and non-nil modified date. On my atomic site (WP.com business plan), a draft product has non-nil creation/modified date. However, in G and Bananza sites, the bug occurs. Feel free to test on any site you prefer, regardless of the draft product API behavior!

Prerequisite: the site has a draft product created/modified earlier than a published product.

- Go to the Products tab
- Tap "Sort by"
- Tap "Date: Newest to Oldest" --> the products should be shown in the same order as in core, especially for the draft product(s)

## Example screenshots

product list in core:

<img width="440" alt="Screen Shot 2020-10-21 at 11 07 11 AM" src="https://user-images.githubusercontent.com/1945542/96680245-b873e000-13a7-11eb-8e79-f3875ed4c79f.png">

before | after
-- | -- 
![Simulator Screen Shot - iPhone 11 - 2020-10-21 at 10 48 07](https://user-images.githubusercontent.com/1945542/96679789-ec023a80-13a6-11eb-933d-0b2905861b43.png) | ![Simulator Screen Shot - iPhone 11 - 2020-10-21 at 11 06 36](https://user-images.githubusercontent.com/1945542/96679791-ef95c180-13a6-11eb-927a-5bc3904755bc.png)
Update release notes:
